mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp

```diff
@@ -199,24 +211,34 @@ LLVMTypeConverter::LLVMTypeConverter(MLIRContext *ctx,
     // original memref type.
     return builder.create<UnrealizedConversionCastOp>(loc, resultType, desc)
         .getResult(0);
-  });
-  // Add generic source and target materializations to handle cases where
-  // non-LLVM types persist after an LLVM conversion.
-  addSourceMaterialization([&](OpBuilder &builder, Type resultType,
-                               ValueRange inputs, Location loc) {
-    if (inputs.size() != 1)
-      return Value();
+  };
 
-    return builder.create<UnrealizedConversionCastOp>(loc, resultType, inputs)
-        .getResult(0);
-  });
+  // Argument materializations convert from the new block argument types
+  // (multiple SSA values that make up a memref descriptor) back to the
+  // original block argument type.
+  addArgumentMaterialization(unrakedMemRefMaterialization);
+  addArgumentMaterialization(rankedMemRefMaterialization);
+  addSourceMaterialization(unrakedMemRefMaterialization);
+  addSourceMaterialization(rankedMemRefMaterialization);
+
+  // Bare pointer -> Packed MemRef descriptor
```

Comment:
"Legality" is a bit of a misnomer in the context of types. Legality is a property of operations (specified via `ConversionTarget`), not types. The type converter converts a type `A` to a type `B`. When when we say "type B is legal", we usually mean that "type B is converted to type B". But we should really say "type B is legal for type A".

Example: In the LLVM type converter, we have the following type conversion rule:
```
ranked memref -> !llvm.struct
```

However, when the bare pointer calling convention is enabled, the `func.func` conversion pattern converts ranked memref block arguments to `!llvm.ptr` block arguments (instead of `!llvm.struct`).

Is `!llvm.ptr` a legal type? Yes, because `!llvm.ptr` converts to `!llvm.ptr`. Is it a legal type for a ranked memref? No, because the type conversion rule states that ranked memrefs should be converted to `!llvm.struct`. And all subsequent lowering patterns expect to receive an `!llvm.struct` from the adaptor. (It's just that the `func.func` conversion pattern chose to convert the memref block argument in a different way.)

> Am I right that this is an issue with the `builtin.unrealized_cast` simplifications? I.e. `!llvm.ptr -> memref -> !llvm.struct` being simplified to `!llvm.ptr -> !llvm.struct`, making this target materialization required?

Not quite. In the case of a bare pointer, there is only one materialization. (There used to be the two materializations that you described.) This is the lowering flow:
1. `func.func` lowering pattern replaces ranked memref block argument with `!llvm.ptr` block argument. No materialization is inserted. That's new. Before this PR, we used to insert an argument materialization and a target materialization. But these materializations are not necessary and increase the complexity of the driver. The argument materialization is not needed because it's a 1:1 replacement. The target materialization is not needed because the lowering pattern of the user of the block argument will insert it.
2. User of memref block argument is lowered. The type converter states that the ranked memref should be converted to `!llvm.struct`, but the pattern finds an `!llvm.ptr` instead. Therefore, it inserts a target materialization from `!llvm.ptr` -> `!llvm.struct`. (This functionality is in `remapValues`.)

Side note: For 1:N replacements (e.g., ranked memref block argument *without* bare pointer calling convention), we still insert an argument materialization in Step 1.

Reply:
> Could you explain why it is desireable to get rid of the argument materailization?

The purpose of an argument materialization is to convert N replacement values into 1 SSA value. It is a workaround around missing 1:N support in the conversion value mapping. In case of a 1:1 conversion, that workaround is not needed because the conversion value mapping can store 1:1 mappings directly.

This is also prefetching a bit from #116524, which removes argument materializations entirely. We are going to need the same type converter setup then.

Reply:
Documentation updated here: https://github.com/llvm/llvm-project/pull/117847

Reply:
> > Could you explain why it is desireable to get rid of the argument materailization?
> 
> The purpose of an argument materialization is to convert N replacement values into 1 SSA value. It is a workaround around missing 1:N support in the conversion value mapping. In case of a 1:1 conversion, that workaround is not needed because the conversion value mapping can store 1:1 mappings directly.
> 
> This is also prefetching a bit from #116524, which removes argument materializations entirely. We are going to need the same type converter setup then.
> 
> Basically I'm trying to slowing phase argument materializations over the next 2-3 PRs.

Doesn't it do more than that though? I was under the impression that argument materialization was a special 1:N source materialization, and that the goal of getting rid of it is to merge its functionality into source materializations. 

Prior to this PR, the code in `DialectConversion.cpp` would also cast to the original SSA value type, both in the 1:1 and 1:N case, (assuming no replacement values), as a source materialization would, which is now no longer the case for 1:1 but is still the case for 1:N (assuming `argument == source materailization`).

Regardless of whether my impression with the source and argument materialization check is correct, it feels odd to me that this code would have different semantics, even if just temporarily, for the 1:1 and 1:N cases.

Reply:
> Doesn't it do more than that though? I was under the impression that argument materialization was a special 1:N source materialization, and that the goal of getting rid of it is to merge its functionality into source materializations.

A source materialization should not be needed at all during `applySignatureConversion` (apart from when block arguments are dropped entirely; but even that could probably be moved to a later point of time). It is only needed when some IR, that still expects the old type, survives the conversion process. That is being taken care of at the end of the conversion process (what used to be "finalize").

The reason why an argument materialization behaves like a source materialization (i.e., back to the original type) is because there's not really any other type it can map to. During `applySignatureConversion` and `replaceOpWithMultiple` we have to decide on an argument materialization result type and the original type is readily available. In theory, any arbitrary type could work, the important thing is that it's a single type that carries enough semantic information for the materialization callbacks to do their work.

> Regardless of whether my impression with the source and argument materialization check is correct, it feels odd to me that this code would have different semantics, even if just temporarily, for the 1:1 and 1:N cases.

I agree, there is an asymmetry and it makes sense to keep argument materialization until we remove it everywhere in one go. I added that code path back to the PR. (Also does not require any LLVM type converter changes any more, as you said.)

Reply:
Actually, this breaks some tests in Flang. There are some patterns that run without a type converter to bypass type conversion. These have caused problems in the past. Maybe it's better to do all of this in the same PR that removes the argument materializations...

Reply:
I thought about this a lot further and I think the initial version you had is probably the best option for now.

Arguably, the issue here with bare pointers is that we have two different type systems, one where memrefs are bare pointers and one where it is the struct. 
Dealing with multiple type systems is complicated and currently done using a target materialization from the type converter of the pattern consuming the value. From that perspective it makes sense that the LLVM type converter would convert a bare pointer whose original type was a memref to the struct.

Regarding the differences in 1:N and 1:1 conversion, I think one could argue that there is no guarantee given to users as to when and why they *must* exist, just in which situations they may exist.

Sorry for the noise :slightly_smiling_face: 

Am I correct that the flang issue now is one of those cases that happened to work and in your original version continue to "happens to work"?

Reply:
> Arguably, the issue here with bare pointers is that we have two different type systems, one where memrefs are bare pointers and one where it is the struct.

Interestingly, in the type converter, there is only one "type system". The LLVM type converter never converts a memref type to an LLVM pointer type. It's a lowering pattern that chooses to replace a memref value with a bare pointer, bypassing the type conversion rules (which is totally fine).

The LLVM type converter and the LLVM lowering patterns were somewhat co-designed. They are used together.

> From that perspective it makes sense that the LLVM type converter would convert a bare pointer whose original type was a memref to the struct.

I think it makes sense for "bare pointer -> struct". I'd say that's pretty "standard" materialization. Both "bare pointer" and "memref descriptor struct" are LLVM types and the two concepts are well documented in the MLIR -> LLVM lowering infrastructure.

Hypothetically speaking, a conversion pattern could replace a memref value with a value of another type, let's say `XYZ`. In that case, the conversion framework would ask the type converter to materialize a target conversion from "`XYZ` -> struct", which will fail. Expectedly. That's not a standard conversion. If users want to support such a conversion in a custom pattern, they can always register additional target materializations.

Not quite sure where I'm going with this. I think there are some "standard" materializations that we can expect the type converter to support out of the box (e.g., "bare ptr -> struct"). But not other stuff such as "`i42` -> struct" (which doesn't even make sense...).

> Am I correct that the flang issue now is one of those cases that happened to work and in your original version continue to "happens to work"?

Yes that's correct. But this Flang issue is a known problem, see #118066 for details. We should not let it affect the design here. Either way, we're either going to fix it or find a temporary workaround in Flang.

I think I'm going to revert this PR back to the original version. And also finalize the `ConversionValueMapping` PR, which is the last missing piece for the 1:1 and 1:N merging. Then we can look at both of them and decide whether we want to merge them separately (with a few days in-between), or squash everything together into one commit.

Reply:
The final dialect conversion PR (#116524) is out for view. Let me know if you prefer separate PRs or one squashed PR.